### PR TITLE
Partial wrapping of pb_ostream_t (pt1)

### PR DESCRIPTION
**Pull request by rsgowman**
Wraps the varint based encode methods.

### CODE REVIEW

**Comment by rsgowman** [comment]
I'm currently about 4 commits into this. See the branch rsgowman/refactor_ostream if you want more context on where this is going. (I can move all those commits into this PR if it would be useful, but will otherwise keep them separate.)

File: Firestore/core/src/firebase/firestore/remote/serializer.cc [hunk 1]
```diff
@@ -39,16 +39,71 @@ void EncodeObject(pb_ostream_t* stream,
...
- * @param value The value to encode, represented as a uint64_t.
+ * Docs TODO(rsgowman). But currently, this just wraps the underlying nanopb
+ * pb_ostream_t. Eventually, this might use static factory methods to create the
+ * underlying pb_ostream_t rather than directly passing it in.
  */
-void EncodeVarint(pb_ostream_t* stream, uint64_t value) {
-  bool status = pb_encode_varint(stream, value);
+class PbOstream {
+ public:
+  explicit PbOstream(pb_ostream_t* stream) : stream_(stream) {
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/remote/serializer.cc]: Do you intend to make this class create the stream on its own later on? I understand why it can't be done in this PR, but it seems to me to be worthwhile.

File: Firestore/core/src/firebase/firestore/remote/serializer.cc [hunk 2]
```diff
@@ -39,16 +39,71 @@ void EncodeObject(pb_ostream_t* stream,
...
- * other Encode* methods (i.e. EncodeBool, EncodeLong, etc)
- *
- * @param value The value to encode, represented as a uint64_t.
+ * Docs TODO(rsgowman). But currently, this just wraps the underlying nanopb
+ * pb_ostream_t. Eventually, this might use static factory methods to create the
+ * underlying pb_ostream_t rather than directly passing it in.
  */
-void EncodeVarint(pb_ostream_t* stream, uint64_t value) {
-  bool status = pb_encode_varint(stream, value);
+class PbOstream {
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/remote/serializer.cc]: Nit: this name just mirrors the Nanopb C-ish convention. I don't have a good suggestion ready, though, so feel free to leave as-is. In particular, I think `Pb` part is unnecessary, and it's a little too abbreviated. `OutputStream`, maybe?

File: Firestore/core/src/firebase/firestore/remote/serializer.cc [hunk 3]
```diff
@@ -39,16 +39,71 @@ void EncodeObject(pb_ostream_t* stream,
...
+void PbOstream::EncodeTag(pb_wire_type_t wiretype, uint32_t field_number) {
+  bool status = pb_encode_tag(stream_, wiretype, field_number);
+  if (!status) {
+    // TODO(rsgowman): figure out error handling
+    abort();
+  }
+}
+
+void PbOstream::EncodeSize(size_t size) {
+  return EncodeVarint(size);
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/remote/serializer.cc]: I presume you `return void` intentionally? (I'm fine with it, just checking)

File: Firestore/core/src/firebase/firestore/remote/serializer.cc [hunk 4]
```diff
@@ -156,59 +211,49 @@ std::string DecodeString(pb_istream_t* stream) {
...
 
     case FieldValue::Type::Integer:
-      status = pb_encode_tag(stream, PB_WT_VARINT,
-                             google_firestore_v1beta1_Value_integer_value_tag);
-      if (!status) {
-        // TODO(rsgowman): figure out error handling
-        abort();
-      }
-      EncodeInteger(stream, field_value.integer_value());
+      stream.EncodeTag(PB_WT_VARINT,
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/remote/serializer.cc]: Can tag encoding be moved into the `Encode*` methods? Or is it realistically possible for a value encoded with `EncodeVarint` to be tagged something other than `PB_WT_VARINT`?

File: Firestore/core/src/firebase/firestore/remote/serializer.cc [hunk 5]
```diff
@@ -39,16 +39,71 @@ void EncodeObject(pb_ostream_t* stream,
...
+   * @param field_number is one of the field tags that nanopb generates based
+   * off of the proto messages. They're typically named in the format:
+   * <parentNameSpace>_<childNameSpace>_<message>_<field>_tag, e.g.
+   * google_firestore_v1beta1_Document_name_tag.
+   */
+  void EncodeTag(pb_wire_type_t wiretype, uint32_t field_number);
+
+  void EncodeSize(size_t size);
+  void EncodeNull();
+  void EncodeBool(bool bool_value);
```
Comment (var-const) [inline Firestore/core/src/firebase/firestore/remote/serializer.cc]: I presume you don't use overloads to avoid unexpected conversions, right? I still think that something like
```cpp
void Encode(bool);
void Encode(int64_t);
template <typename T>
void Encode(T) = delete;
```
might be interesting, because it could open the door to making code more generic. If there's no opportunity for that, though, then it's not worth it.

Comment (rsgowman) [reply Firestore/core/src/firebase/firestore/remote/serializer.cc]: Since this is kinda an 'ostream' like thing, you could even replace Encode* with operator<<. :)

For now, I've left this as being explicit. (I believe we're trying to err on the side of too explicit for internal apis.)

File: Firestore/core/src/firebase/firestore/remote/serializer.cc (hunk 3 above)
Comment (rsgowman) [reply Firestore/core/src/firebase/firestore/remote/serializer.cc]: Yes. Eventually, void->error_code throughout. (probably.)

File: Firestore/core/src/firebase/firestore/remote/serializer.cc (hunk 4 above)
Comment (rsgowman) [reply Firestore/core/src/firebase/firestore/remote/serializer.cc]: It could. (In fact, it used to be.) But we pulled it back out due to an asymmetry with the decode methods.

tl;dr for the decode is that while EncodeVarint/Bool/Int all use PB_WT_VARINT (so mapping is easy) but the reversing it is tricky since it's difficult to know if a PB_WT_VARINT should be decoded as a varint/bool/int. (EncodeString/Submessage also have the same problem as they both map to PB_WT_STRING.)

This is something we can take another look at, but it's probably easiest to do that after the rest of the refactoring (incl decoding) is done.

File: Firestore/core/src/firebase/firestore/remote/serializer.cc (hunk 2 above)
Comment (var-const) [reply Firestore/core/src/firebase/firestore/remote/serializer.cc]: In my view, the protobuf part is kinda implied because it's in `serializer`, but I don't feel strongly.

File: Firestore/core/src/firebase/firestore/remote/serializer.cc (hunk 4 above)
Comment (var-const) [reply Firestore/core/src/firebase/firestore/remote/serializer.cc]: Agreed, let's revisit further on.

File: Firestore/core/src/firebase/firestore/remote/serializer.cc (hunk 2 above)
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/remote/serializer.cc]: `WriteStream`? `Writer`?

This type is entirely local to the serializer so I don't think there's a whole lot of benefit to being super verbose here.

File: Firestore/core/src/firebase/firestore/remote/serializer.cc (hunk 5 above)
Comment (wilhuff) [reply Firestore/core/src/firebase/firestore/remote/serializer.cc]: Please don't make this a std::ostream-like thing. We don't want any ADL in here nor do we want anything defined specifically for ostream to get involved.

We can push to make this more generic when we need it, but likely YAGNI.

File: Firestore/core/src/firebase/firestore/remote/serializer.cc [hunk 6]
```diff
@@ -39,16 +39,71 @@ void EncodeObject(pb_ostream_t* stream,
...
+    abort();
+  }
+}
+
+void PbOstream::EncodeSize(size_t size) {
+  return EncodeVarint(size);
+}
+
+void PbOstream::EncodeVarint(uint64_t value) {
+  bool status = pb_encode_varint(stream_, value);
```
Comment (wilhuff) [inline Firestore/core/src/firebase/firestore/remote/serializer.cc]: Just to be sure: error handling changes are coming subsequent to this PR, yes?

File: Firestore/core/src/firebase/firestore/remote/serializer.cc [hunk 7]
```diff
@@ -291,7 +336,7 @@ void EncodeNestedFieldValue(pb_ostream_t* stream,
   size_t size = substream.bytes_written;
 
   // Write out the size to the output stream.
-  EncodeVarint(stream, size);
+  PbOstream(stream).EncodeSize(size);
```
Comment (wilhuff) [inline Firestore/core/src/firebase/firestore/remote/serializer.cc]: This may just be a hot take on my part, but it seems that your abstraction over `pb_ostream_t` is incomplete. What if instead of taking a paramter of `pb_ostream_t*` you took a `PbOstream*` (or whatever you call it)?

File: Firestore/core/src/firebase/firestore/remote/serializer.cc (hunk 2 above)
Comment (var-const) [reply Firestore/core/src/firebase/firestore/remote/serializer.cc]: FWIW, I like `Writer`.

Comment (rsgowman) [reply Firestore/core/src/firebase/firestore/remote/serializer.cc]: Suggestions are:
- PbOstream, OutputStream, WriteStream, Writer

I'll keep the 'Stream' part, since nobody seems to hate that.

Output vs Writer: I don't care. I don't obviously prefer Writer, so I guess I'll stick with Output. (If you prefer Writer to Output, i.e. you weren't just floating another option, then I'll use Writer.)

Final name (for now): OutputStream.

Comment (rsgowman) [reply Firestore/core/src/firebase/firestore/remote/serializer.cc]: Update: Costa likes Writer. That's two votes for Writer, so I'll use that.

File: Firestore/core/src/firebase/firestore/remote/serializer.cc [hunk 8]
```diff
@@ -39,16 +39,71 @@ void EncodeObject(pb_ostream_t* stream,
...
+   * Encodes a message type to the output stream.
+   *
+   * This essentially wraps calls to nanopb's pb_encode_tag() method.
+   *
+   * @param field_number is one of the field tags that nanopb generates based
+   * off of the proto messages. They're typically named in the format:
+   * <parentNameSpace>_<childNameSpace>_<message>_<field>_tag, e.g.
+   * google_firestore_v1beta1_Document_name_tag.
+   */
+  void EncodeTag(pb_wire_type_t wiretype, uint32_t field_number);
```
Comment (wilhuff) [inline Firestore/core/src/firebase/firestore/remote/serializer.cc]: I know this is _super_ annoying but I would expect that writers write, not encode.

Feel free to push this off with a todo or whatever until we've progressed down more of your review stack.

Comment (rsgowman) [reply Firestore/core/src/firebase/firestore/remote/serializer.cc]: Added TODO.
